### ISSUE TITLE
One Shot Learning and Siamese Networks in Keras

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/readinglist#305]
https://sorenbouma.github.io/blog/oneshot/

**Comment by sweaterr** [comment]
## Background:

기존의 지혜에 따르면 심 신경 네트워크는 이미지 나 음성 언어와 같은 고차원 데이터를 통해 학습하는 것이 매우 훌륭하지만, 훈련을 위한 엄청난 양의 예제가 있을 때만 가능합니다. 반면에 인간은 원샷 학습이 가능합니다. 이전에 주걱을 본 적이 없는 사람을 데리고 주걱을 한 장의 그림으로 보여 주면 높은 정밀도로 다른 주걱과 구별 할 수 있습니다 

Yet another one of the things humans can do that seemed trivial to us right up until we tried to make an algorithm do it. 아주 적은 양의 데이터로 신속하게 학습 할 수 있는 능력은 데이터를 수집하고 라벨링하는 것이 값 비싸기 때문에 기계 학습 시스템에서 분명히 바람직한 것처럼 보입니다. 나는 또한 이것이 일반적인 지능에 대한 긴 길에서 중요한 단계라고 생각한다. 최근에는 신경망을 이용한 원샷 학습에 대한 많은 흥미로운 논문들이 있었으며 좋은 결과를 얻었습니다. 이것은 정말 흥미로운 새로운 영역이므로 동료 신입생이 깊은 학습에 보다 쉽게 ​​접근 할 수 있도록 쉬운 소개를 하고 싶었습니다.

**Comment by sweaterr** [comment]
## Formulating the Problem - N-way One-Shot Learning
Before we try to solve any problem, we should first precisely state what the problem actually is, so here is the problem of one-shot classification expressed symbolically:  Our model is given a tiny labelled training set \( S \), which has N examples, each vectors of the same dimension with a distinct label \(y\).
어떤 문제를 풀기 전에 먼저 문제가 실제로 무엇인지를 명시해야합니다. 따라서 여기에 원샷 분류의 문제가 수학적으로 표현되어야 합니다. 우리 모델에는 작은 레이블이 지정된 훈련 집합 $S$이 주어지며 N 개의 예제가 있습니다.

$S = \{ (x_1,y_1), …, (x_N,y_N) \}$

$\hat {x}$도 제공됩니다. 이 예제는 분류해야하는 테스트 예제입니다. Since exactly one example in the support set has the right class, the aim is to correctly predict which $ y \in S $is the same as $ \hat{x} $ ‘s label, $\hat{y}$. 

문제를 정의하는 좋은 방법이 있지만, 여기서는 다루지 않는다. 다음을 기억해 둬라
* 실세계 문제는 항상 하나의 이미지가 정답이라는 제약을 갖는 것은 아니다
* k-shot으로 일반화하기 쉽다.
* N이 크면 문제가 더 어려워진다.
* 랜덤 추측은 $\frac{100}{n} \%$ 정확도를 가진다.

**Comment by sweaterr** [comment]
## 데이터에 대해서  - Omniglot! :

Omniglot 데이터 세트는 50 개의 알파벳으로 된 1623 개의 손으로 그린 ​​문자의 모음입니다. 모든 캐릭터에는 해상도가 105x105 인 다른 사람이 그린 각각 20 개의 예가 있습니다.

기계 학습을 좋아한다면 MNIST 데이터 세트에 대해 들어 보셨을 것입니다.

Omniglot은 mnist의 전치 (transpose)라고도하며 MNIST와는 달리 1623개의 문자가 각각 20 개 예제 밖에 없습니다. 각 캐릭터를 만드는 데 사용 된 스트로크에 대한 데이터도 있지만, 우리는 이를 사용하지 않을 것입니다. 보통 30 개의 훈련 알파벳과 20 개의 평가 알파벳으로 나뉩니다. 이러한 모든 다른 문자로 인해 가능한 많은 원샷 작업이 이루어지므로 원샷 학습 알고리즘에 대한 벤치 마크는 매우 좋습니다.

**Comment by sweaterr** [comment]
## A One-Shot Learning Baseline / 1 Nearest Neighbour

The simplest way of doing classification is with k-nearest neighbours, but since there is only one example per class we have to do 1 nearest neighbour.
분류를 수행하는 가장 간단한 방법은 k- 가장 가까운 이웃을 사용하는 것이지만, 클래스 당 하나의 예가 있기 때문에 가장 가까운 이웃을 수행해야합니다.

This is very simple, just calculate the Euclidean distance of the test example from each training example and pick the closest one:.
이것은 매우 간단합니다. 각 예제에서 유클리드 거리를 계산하고 가장 가까운 것을 선택하십시오.

**Comment by sweaterr** [comment]
## Ways to use deep networks for one shot learning?!

**Comment by sweaterr** [comment]
## Siamese networks

**Comment by sweaterr** [comment]
## Network architecture
Unfortunately, properly explaining how and why a convolutional neural net work would make this post twice as long.
불행히도, 어떻게 왜 왜곡 된 신경 네트워크 작업이이 게시물을 두 배나 길게 만들 것이라고 설명합니다.

If you want to understand convnets work, I suggest checking out cs231n and then colah.
만약 당신이 convnets 작업을 이해하고 싶다면, 나는 cs231n과 colah를 체크 아웃 할 것을 권한다.

For any non-dl people who are reading this, the best summary I can give of a CNN is this: An image is a 3D array of pixels.
이것을 읽지 않는 비 dl 사람들은 CNN에 대한 가장 좋은 요약은 다음과 같습니다. 이미지는 픽셀의 3D 배열입니다.

A convolutional layer is where you have a neuron connected to a tiny subgrid of pixels or neurons, and use copies of that neuron across all parts of the image/block to make another 3d array of neuron activations.
길쌈 레이어 (convolutional layer)는 픽셀 또는 뉴런의 작은 서브 그리드에 연결된 뉴런을 가지고 있으며, 이미지 / 블록의 모든 부분에 걸쳐 해당 뉴런의 복사본을 사용하여 다른 3 차원 뉴런 배열을 만듭니다.

A max pooling layer makes a block of activations spatially smaller.
최대 풀링 계층은 활성화 블록을 공간적으로 더 작게 만듭니다.

Lots of these stacked on top of one another can be trained with gradient descent and are really good at learning from images.
서로 위에 쌓인이 많은 것들은 그라디언트 강하로 훈련 될 수 있고 이미지로부터 배우는 것에 능숙합니다.

I’m going to describe the architecture pretty briefly because it’s not the important part of the paper.
이 논문의 중요한 부분이 아니기 때문에이 아키텍처를 아주 간략하게 설명 할 것입니다.

Koch et al uses a convolutional siamese network to classify pairs of omniglot images, so the twin networks are both convolutional neural nets(CNNs).
코흐 외 (Koch et al)는 옴니 슬롯 이미지 쌍을 분류하기 위해 길쌈 (convolutional) siamese 네트워크를 사용하므로 트윈 네트워크는 모두 CNN (convolutional neural net)이다.

The twins each have the following architecture: convolution with 64 10x10 filters, relu -> max pool -> convolution with 128 7x7 filters, relu -> max pool -> convolution with 128 4x4 filters, relu -> max pool -> convolution with 256 4x4 filters.
쌍둥이는 각각 다음과 같은 아키텍처를 가지고 있습니다 : 64 개의 10x10 필터를 가진 컨볼 루션, 128 개의 7x7 필터를 가진 relu -> 최대 풀 -> 컨볼 루션, 128 개의 4x4 필터를 가진 relu -> 최대 풀 -> 컨볼 루션, relu -> 최대 풀 -> 256 번 컨볼 루션 4x4 필터.

The twin networks reduce their inputs down to smaller and smaller 3d tensors, finally their is a fully connected layer with 4096 units.
트윈 네트워크는 작고 작은 3D 텐서로 입력을 줄이고 마지막으로 4096 단위의 완전히 연결된 레이어입니다.

The absolute difference between the two vectors is used as input to a linear classifier.
두 벡터의 절대 차는 선형 분류 자에 대한 입력으로 사용됩니다.

All up, the network has 38,951,745 parameters - 96% of which belong to the fully connected layer.
모두 위로, 네트워크는 38,951,745 개의 매개 변수를 가지고 있습니다. 그 중 96 %는 완전히 연결된 레이어에 속합니다.

This is quite a lot, so the network has high capacity to overfit, but as I show below, pairwse training means the dataset size is huge so this won’t be a problem.
이것은 상당히 많기 때문에 네트워크의 용량이 많습니다.하지만 아래에 나와있는 것처럼 쌍방향 교육은 데이터 집합 크기가 너무 커서 문제가되지 않을 것입니다.

**Comment by sweaterr** [comment]
## Observation: effective dataset size in pairwise training

EDIT: After discussing this with a PhD student at UoA, I think this bit might be overstated or even just wrong. Emperically, my implementation did overfit, even though it wasn’t trained for enough iterations to sample every possible pair, which kind of contradicts this section. I’m leaving it up in the spirit of being wrong loudly.

One cool thing I noticed about training on pairs is that there are quadratically many possible pairs of images to train the model on, making it hard to overfit. Say we have $ C$ examples each of $ E $ classes. Since there are $ C \cdot E $ images total, the total number of possible pairs is given by
Npairs=(C⋅E2)=(C⋅E)!2!(C⋅E−2)!
For omniglot with its 20 examples of 964 training classes, this leads to 185,849,560 possible pairs, which is huge! However, the siamese network needs examples of both same and different class pairs. There are $ E$ examples per class, so there will be $ {\binom {E}{2}} $ pairs for every class, which means there are $N_{same} = {\binom {E}{2}} \cdot C $ possible pairs with the same class - 183,160 pairs for omniglot. Even though 183,160 example pairs is plenty, it’s only a thousandth of the possible pairs, and the number of same-class pairs increases quadratically with E but only linearly with C. This is important because the siamese network should be given a 1:1 ratio of same-class and different-class pairs to train on - perhaps it implies that pairwise training is easier on datasets with lots of examples per class.
